### PR TITLE
typecheck: Add basic typechecking for ReferenceType

### PR DESCRIPTION
**Pull request by goar5670**
provide basic resolve and typechecking for `ReferencePattern`.

### CODE REVIEW

**Comment by powerboat9** [comment]
@goar5670 the clang format check failed

File: gcc/rust/typecheck/rust-hir-type-check-pattern.cc [hunk 1]
```diff
@@ -410,8 +410,13 @@ TypeCheckPattern::visit (HIR::QualifiedPathInExpression &pattern)
 void
 TypeCheckPattern::visit (HIR::ReferencePattern &pattern)
 {
-  rust_sorry_at (pattern.get_locus (),
-		 "type checking qualified path patterns not supported");
+  if (parent->get_kind () != TyTy::TypeKind::REF)
+    rust_error_at (pattern.get_locus (), "expected %s, found reference",
```
Comment (philberty) [inline gcc/rust/typecheck/rust-hir-type-check-pattern.cc:414]: You could add a dg-error test case for this

**Review by philberty — CHANGES_REQUESTED** [review]
Seems good to me but lets add a test case for this

File: gcc/rust/typecheck/rust-hir-type-check-pattern.cc (hunk 1 above)
Comment (goar5670) [reply gcc/rust/typecheck/rust-hir-type-check-pattern.cc:414]: Done 👍 please recheck

**Review by philberty — APPROVED** [review]
LGTM

**Pull request merged**